src/gui/entry/EntryAttachmentsWidget.cpp

```diff
@@ -0,0 +1,358 @@
+            return;
+        }
+    }
+    config()->set("LastAttachmentDir", QFileInfo(saveDir.absolutePath()).absolutePath());
+
+    QStringList errors;
+    for (const QModelIndex &index: indexes) {
+        const QString filename = m_attachmentsModel->keyByIndex(index);
+        const QString attachmentPath = saveDir.absoluteFilePath(filename);
+
+        if (QFileInfo::exists(attachmentPath)) {
+            const QString question(tr("Are you sure you want to overwrite existing file \"%1\" with the attachment?"));
+            auto answer = MessageBox::question(this, tr("Confirm overwrite"), question.arg(filename),
+                                               QMessageBox::Yes | QMessageBox::No | QMessageBox::Cancel);
+            if (answer == QMessageBox::No) {
+                continue;
+            } else if (answer == QMessageBox::Cancel) {
+                return;
+            }
+        }
+
+        QFile file(attachmentPath);
+        const QByteArray attachmentData = m_entryAttachments->value(filename);
+        const bool saveOk = file.open(QIODevice::WriteOnly) && file.write(attachmentData) == attachmentData.size();
+        if (!saveOk) {
+            errors.append(QString("%1 - %2").arg(filename, file.errorString()));
+        }
+    }
+
+    if (!errors.isEmpty()) {
+        errorOccurred(tr("Unable to save the attachments:\n").append(errors.join('\n')));
+    }
+}
+
+void EntryAttachmentsWidget::openAttachment(const QModelIndex& index)
+{
+    Q_ASSERT(index.isValid());
+    if (!index.isValid()) {
+        return;
+    }
+
+    QString errorMessage;
+    if (!openAttachment(index, &errorMessage)) {
+        errorOccurred(tr("Unable to open the attachment:\n%1").arg(errorMessage));
+    }
+}
+
+void EntryAttachmentsWidget::openSelectedAttachments()
+{
+    const QModelIndexList indexes = m_ui->attachmentsView->selectionModel()->selectedRows(0);
+    if (indexes.isEmpty()) {
+        return;
+    }
+
+    QStringList errors;
+    for (const QModelIndex &index: indexes) {
+        QString errorMessage;
+        if (!openAttachment(index, &errorMessage)) {
+            const QString filename = m_attachmentsModel->keyByIndex(index);
+            errors.append(QString("%1 - %2").arg(filename, errorMessage));
+        };
+    }
+
+    if (!errors.isEmpty()) {
+        errorOccurred(tr("Unable to open the attachments:\n").append(errors.join('\n')));
+    }
+}
+
+void EntryAttachmentsWidget::updateButtonsEnabled()
+{
+    const bool hasSelection = m_ui->attachmentsView->selectionModel()->hasSelection();
+
+    m_ui->addAttachmentButton->setEnabled(!m_readOnly);
+    m_ui->removeAttachmentButton->setEnabled(hasSelection && !m_readOnly);
+
+    m_ui->saveAttachmentButton->setEnabled(hasSelection);
+    m_ui->openAttachmentButton->setEnabled(hasSelection);
+}
+
+bool EntryAttachmentsWidget::insertAttachments(const QStringList& filenames, QString* errorMessage)
```

Comment:
Why pointer to QString instead of non-const reference?

Reply:
In my opinion, this implementation is more flexible and more expressive.
Pros:
- the caller obviously has to pass a reference that explicitly indicates that the passed argument can be changed
- in case we do not need a error message, we just send nullptr 

For example:

```cpp
func1(param, result);
// vs 
func2(param, &result);
```
In case with func2, we do not need to look at the function declaration to understand that the second parameter can be changed in the function

Reply:
A non-const reference is already an explicit indication that the value will be passed back. In case of no error message, it just remains unchanged.
If you use a pointer, you need to explicitly check for nullptr before you dereference it inside the function, which is quite unnecessary. Passing in a pointer also requires you to allocate a string on the freestore for no apparent reason (unless you really want to go down the rabbit hole of passing pointers to stack variables to the function).